subsys/input/input_utils.c

```diff
@@ -111,6 +115,148 @@ static int input_cmd_report(const struct shell *sh, size_t argc, char *argv[])
+
+/* Keep space for each column value, 2 char per byte + space. */
+#define KEY_MATRIX_ENTRY_LEN (sizeof(kbd_row_t) * 2 + 1)
+#define KEY_MATRIX_BUF_SZ (CONFIG_INPUT_KBD_MATRIX_STATE_MAX_COLS * \
+			   KEY_MATRIX_ENTRY_LEN)
+static char kbd_matrix_buf[KEY_MATRIX_BUF_SZ];
+
+static void kbd_matrix_state_log_entry(char *header, kbd_row_t *data)
+{
+	const struct input_kbd_matrix_common_config *cfg = kbd_matrix_state_dev->config;
+	char *buf = kbd_matrix_buf;
+	int size = sizeof(kbd_matrix_buf);
+	int ret;
+	char blank[KEY_MATRIX_ENTRY_LEN];
+	int count = 0;
+
+	memset(blank, '-', sizeof(blank) - 1);
+	blank[sizeof(blank) - 1] = '\0';
+
+	for (int i = 0; i < cfg->col_size; i++) {
+		char *sep = (i + 1) < cfg->col_size ? " " : "";
+
+		if (data[i] != 0) {
+			ret = snprintf(buf, size, PRIkbdrow "%s", data[i], sep);
+		} else {
+			ret = snprintf(buf, size, "%s%s", blank, sep);
+		}
+		size -= ret;
+		buf += ret;
+
+		count += POPCOUNT(data[i]);
+
+		/* Last byte is for the string termination */
+		if (size < 1) {
+			LOG_ERR("kbd_matrix_buf too small");
+			return;
+		}
+	}
+
+	LOG_INF("%s %s [%s] (%d)",
+		kbd_matrix_state_dev->name, header, kbd_matrix_buf, count);
+}
+
+static void kbd_matrix_state_log(struct input_event *evt)
+{
+	const struct input_kbd_matrix_common_config *cfg;
+	static int row, col, val;
+
+	if (kbd_matrix_state_dev == NULL || kbd_matrix_state_dev != evt->dev) {
+		return;
+	}
+
+	cfg = kbd_matrix_state_dev->config;
+
+	switch (evt->code) {
+	case INPUT_ABS_X:
+		col = evt->value;
+		break;
+	case INPUT_ABS_Y:
+		row = evt->value;
+		break;
+	case INPUT_BTN_TOUCH:
+		val = evt->value;
+		break;
+	}
+
+	if (!evt->sync) {
+		return;
+	}
+
+	if (col > (CONFIG_INPUT_KBD_MATRIX_STATE_MAX_COLS - 1)) {
+		LOG_ERR("column index too large for the state buffer: %d", col);
+		return;
+	}
+
+	if (col > (cfg->col_size - 1)) {
+		LOG_ERR("invalid column index: %d", col);
+		return;
+	}
+
+	if (row > (cfg->row_size - 1)) {
+		LOG_ERR("invalid row index: %d", row);
+		return;
+	}
+
+	WRITE_BIT(kbd_matrix_state[col], row, val);
+	if (val != 0) {
+		WRITE_BIT(kbd_matrix_key_mask[col], row, 1);
+	}
+
+	kbd_matrix_state_log_entry("state", kbd_matrix_state);
+}
+INPUT_CALLBACK_DEFINE(NULL, kbd_matrix_state_log);
+
+static int input_cmd_kbd_matrix_state(const struct shell *sh,
+				      size_t argc, char *argv[])
+{
+	const struct device *dev;
+
+	if (!strcmp(argv[1], "off")) {
+		if (kbd_matrix_state_dev != NULL) {
+			kbd_matrix_state_log_entry("key-mask",
+						   kbd_matrix_key_mask);
+		}
+
+		kbd_matrix_state_dev = NULL;
+		shell_info(sh, "Keyboard state logging disabled");
+		return 0;
+	}
+
+	dev = device_get_binding(argv[1]);
+	if (dev == NULL) {
+		shell_error(sh, "Invalid device: %s", argv[1]);
+		return -ENODEV;
+	}
+
+	memset(kbd_matrix_state, 0, sizeof(kbd_matrix_state));
+	memset(kbd_matrix_key_mask, 0, sizeof(kbd_matrix_state));
+	kbd_matrix_state_dev = dev;
```

Comment:
In the unlikely event that there are 2 keyboard matrix input devices, it could make sense to check that `kbd_matrix_state_dev` is NULL.

Reply:
good idea, added a check to block changing a device without passing through the "off" state